lecture-notes/2-vcs.md

```diff
@@ -1940,37 +1942,102 @@ After you have rebased, you can safely push your (renewed) local commits to the
+> things: rewriting history is fine if it's a private, short-lived development
+> branch. Otherwise, merge.
 
 #### Merge
 
 It's also possible to use `git merge` to reconcile local and remote changes.
-Merge also comes with ability to fast-forward but that must be requested by
-default with `git merge --ff`.
+Like rebase, merge also comes with ability to fast-forward, but that does not
+happen automatically; it must be requested with `git merge --ff` or
+`--ff-only`.
+
+We'll talk about `--ff-only` first because it's the most similar to what you
+have seen so far. If you run `git merge --ff-only origin/branch-name`, the
+merge will only succeed if it need not create a merge commit. That is, if the
+local branch can be fast-forwarded to the remote, great. If not, noisily fail.
 
-First case: `--ff-only`
+In the case where the two branches diverge (`--ff-only` would fail), you can
+use `git merge origin/branch-name` to make a merge commit between your local
+branch and the remote-tracking branch.
+
+A middle ground is `git merge --ff`, which will try to fast-forward if it can,
+and otherwise fall back to creating a merge commit.
 
 #### Pull
 
-TODO: update your local branches via `git merge --ff-only` or `git rebase`.
-Mention `git pull` is a shortcut for fetch+merge but can be unintuitive.
+Now, we've talked about running `git fetch` and then either `git rebase` or
+`git merge` to reconcile differences between local and remote commits. Running
+two commands each time can be cumbersome, so Git provides a shortcut. We
+mention the shortcut last for two reasons:
 
-TODO: Deleting branches with `git push :foo`
+1. We want you to understand what is going behind the scenes; this combo
+   command is not magical
+2. The combo command can behave in unexpected ways, especially if you don't
+   understand the underlying concepts
+
+The command is `git pull`. Now, directly from the man page:
+
+> Incorporates changes from a remote repository into the current branch. If the
+> current branch is behind the remote, then by default it will fast-forward the
+> current branch to match the remote. If the current branch and the remote have
+> diverged, the user needs to specify how to reconcile the divergent branches
+> with `--rebase` or `--no-rebase` [...].
+
+> More precisely, `git pull` runs `git fetch` with the given parameters and
+> then depending on configuration options or command line flags, will call
+> either `git rebase` or `git merge` to reconcile diverging branches.
+
+Now that you have both learned about remote tracking branches and refreshed
+your memory about rebase/merge, this may sound like absolutely no big deal. Of
+course that's what it does, right? Why would it do anything else?
+
+Unfortunately, most people who start off with Git start with a
+`commit`+`push`+`pull` workflow and don't understand why things break when they
+do, or don't understand why there are these funky merge commits sometimes. Now
+you know.
```

Comment:
I'd elaborate more on what people's specific misconceptions tend to be, if you're going to mention it at all.

Reply:
"Unfortunately, many people learn `pull` (and `push`) only as a black box that magically synchronizes a local branch to a remote one (or vice versa). Since much of the commands' behavior can only be explained by knowing the separate steps they take, those people risk an unpleasant surprise in all but the most simple branch states."